micrometer/src/main/scala/zio/metrics/connectors/micrometer/MicrometerMetricListener.scala

```diff
@@ -0,0 +1,89 @@
+package zio.metrics.connectors.micrometer
+
+import java.time.Instant
+import java.util.concurrent.{ConcurrentHashMap, ConcurrentMap}
+
+import scala.jdk.CollectionConverters._
+
+import zio.{Unsafe, URIO, ZIO}
+import zio.metrics.{MetricKey, MetricKeyType, MetricListener}
+import zio.metrics.MetricKeyType.{Counter, Frequency, Gauge}
+import zio.metrics.connectors.micrometer.internal.AtomicDouble
+
+import io.micrometer.core.instrument.{DistributionSummary, MeterRegistry, Tag}
+
+private[micrometer] class MicrometerMetricListener(
+  meterRegistry: MeterRegistry,
+  config: MicrometerConfig,
+  activeGauges: ConcurrentMap[MetricKey.Gauge, AtomicDouble])
+    extends MetricListener {
+  private def getOrCreateGaugeRef(key: MetricKey[Gauge]): AtomicDouble =
+    activeGauges.computeIfAbsent(
+      key,
+      _ =>
+        meterRegistry.gauge(
+          key.name,
+          micrometerTags(key.tags).asJava,
+          AtomicDouble.make(0),
+          (v: AtomicDouble) => v.get(),
+        ),
+    )
+
+  override def updateHistogram(
+    key: MetricKey[MetricKeyType.Histogram],
+    value: Double,
+  )(implicit unsafe: Unsafe,
+  ): Unit =
+    DistributionSummary
+      .builder(key.name)
+      .tags(micrometerTags(key.tags).asJava)
+      .serviceLevelObjectives(key.keyType.boundaries.values.filter(_ > 0): _*) // micrometer prohibits <= 0 slo values
+      .register(meterRegistry)
+      .record(value)
+
+  override def updateGauge(key: MetricKey[Gauge], value: Double)(implicit unsafe: Unsafe): Unit =
+    getOrCreateGaugeRef(key).set(value)
+
+  override def modifyGauge(key: MetricKey[Gauge], value: Double)(implicit unsafe: Unsafe): Unit =
+    getOrCreateGaugeRef(key).incrementBy(value)
+
+  override def updateFrequency(key: MetricKey[Frequency], value: String)(implicit unsafe: Unsafe): Unit =
+    meterRegistry
+      .counter(
+        key.name,
+        (micrometerTags(key.tags) ++ Iterable(Tag.of("bucket", value))).asJava,
+      )
+      .increment()
+
+  override def updateSummary(
+    key: MetricKey[MetricKeyType.Summary],
+    value: Double,
+    instant: Instant,
+  )(implicit unsafe: Unsafe,
+  ): Unit =
+    DistributionSummary
+      .builder(key.name)
+      .tags(
+        (micrometerTags(key.tags) ++ Iterable(Tag.of("error", key.keyType.error.toString))).asJava,
+      )
+      .distributionStatisticBufferLength(key.keyType.maxSize)
+      .distributionStatisticExpiry(key.keyType.maxAge)
+      .publishPercentiles(key.keyType.quantiles: _*)
+      .percentilePrecision(config.summaryPercentileDigitsOfPrecision)
+      .register(meterRegistry)
+      .record(value)
```

Comment:
I think we could improve the performance of this implementation if we would not "build" the metrics each time. Similar to what you did with the gauges, we could have a map of `MetricKey.Summary` to `DistributionSummary` instances that are already registered with micrometer, and the same for all metric types.  I think that should be at least 2 times faster.